### DISCUSSION TITLE
Set a custom size for the usable area

### DISCUSSION THREAD

**Discussion by SeanPluemer** [rdudhagra/Sand-Table#16] [Q&A] [answered]
Is there any way that we would be able to input a custom size for the area?
The reason I ask this is that I have a non-square shape. You can see from my attached video, the gantry will home at the bottom left, which will be the reference 0,0 point. Is there any way that I can home, and then set the 0,0 to a custom location? 
Also, is there any way to make a custom shape with gcode, as you can see in the video, my usable area is an irregular shape. 

https://user-images.githubusercontent.com/58951141/114428292-3ce63500-9b79-11eb-832d-b11acf8c0125.mp4

**Comment by rdudhagra** [comment, marked as answer]
### **Is there any way that we would be able to input a custom size for the area?**
You can set the rectangular bounds of the sand table's size in the `.env` file (as well as in the Marlin firmware, you must do both). However, I assume what you want here is for the website to "crop" the uploaded track to fit in the confines of your usable area. Currently, I'm cropping to a rectangular shape using a series of `if` statements; but it seems that your case would require something more complicated. You'd have to edit `process_theta_rho.js`, specifically the `scale()` function, to work with a more complex shape, perhaps a polyline defined by a series of points that roughly matches the usable area. Take a look at [this NPM package](https://www.npmjs.com/package/point-in-polygon)—this should solve half the issue...the other half will be trimming the out-of-bounds coordinate back to the bounds of the defined polyline (trivial for a rectangle, more difficult for non-standard shapes).

### **The reason I ask this is that I have a non-square shape. You can see from my attached video, the gantry will home at the bottom left, which will be the reference 0,0 point. Is there any way that I can home, and then set the 0,0 to a custom location?**
Yep! Take a look at the end of `gcode_controller.js` and find the code block that looks like this:
```javascript
setTimeout(() => {
    // Start gcode
    backwards = false;
    index = 0;
    file = [
        "G90",
        "G28",
        "M220 S100",
        "M204 P4000 T4000",
        "M205 J0.0001",
        `G0 X${Process_Theta_Rho.X_SIZE / 2} Y${Process_Theta_Rho.Y_SIZE / 2}`, // center of table
    ];
    sendNextCmd(true); // Run start gcode
    console.log("GCode Controller Initialized");
}, 5000);
```
This is the "start gcode" that runs when the table is first powered up. As you can see, it first homes, sets some speed/jerk parameters, and then navigates to the center of the table. Right after the `"G28"` line, add the following two lines:
```javascript
"G0 X[The X offset from the current position to the desired (0,0)] Y[The Y offset from the current position to the desired (0,0)]",
"G92 X0 Y0",
```
The first line will navigate to where you want the new origin to be, and the second will set that position as the new (0,0). Keep in mind that your bounds (max X, max Y in `.env`) will need to be relative to this new origin, not the old one.

### **Also, is there any way to make a custom shape with gcode, as you can see in the video, my usable area is an irregular shape.**
Not sure what you mean here...you could potentially use Inkscape to draw the shape you want as an `.svg` file and then convert that to gcode? Or did I answer this part already somewhere above?

Let me know if this answers your question...and good luck with your sand table (It's looking great so far)!

**Reply by SeanPluemer** [reply]
Thanks as always for the quick resonses!

1. I never even thought about cropping, that's a good idea.
2. The custom home point looks straight enough.
3. Tbh, I don't even think I knew what I was writing for the 3rd point. Ignore that for now.

Moving forward, I think the cropping idea is pretty straightforward. We will see how hard this is in practice. 
One question, you said, "the other half will be trimming the out-of-bounds coordinate back to the bounds of the defined polyline (trivial for a rectangle, more difficult for non-standard shapes)." 
What exactly do you mean by this?

My understanding of how the code should work is this:
1. Import thr file.
2. Raspberry pi converts thr to gcode, based on the size of the table.
3. Scale function is called to crop that gcode into the special shapes? (by moving out of bounds points into the edge of ur crop)

I will check the code when I get some free time, as you can tell... I am moving super slowly on this project!

**Reply by rdudhagra** [reply]
Haha nobody really has enough free time these days :(

Yes, you are 100% correct as to how the scale function works. The current code will automatically crop the design to the bounds specified in `.env`...an easy way to see this/play around with it is to just set the bounds and upload some tracks...you should be able to visualize the cropping in action on the website. Currently, there are two modes: fit (shrinks the `.thr` to fit the smallest dimension of the rectangular bounds) and fill (which will enlarge the `.thr` to fit the largest dimension, cropping the sides off).

 So the challenge is that if for instance, the command tells the gantry to move past the bounds, the software must (1) allow the gantry to move normally until it hits the bounds, (2) then move along the bounds until it gets to as close as possible to the intended destination. This is what should happen for rectangular bounds:
![image](https://user-images.githubusercontent.com/14984724/114450186-b2ff9180-9ba3-11eb-8e17-9c1fea158be7.png)

However, I make a simple approximation that makes the calculation much easier. The way that I process the `.thr` to gcode is by splitting it up into very small steps and approximating those arcs as lines. By doing this, we can simply just clamp the starting and ending coordinate and not worry about the above complexity, and our result will be "close enough". For rectangles, it's as simple as clamping the coordinate between 0 and the max for that axis. In the image below, the yellow arrows get remapped to the red arrows that point vertically, which is the result of doing the above. 
![image](https://user-images.githubusercontent.com/14984724/114451147-e989dc00-9ba4-11eb-8333-007d0ecdad80.png)

What I thought you meant was potentially cropping to shapes that are not simple rectangles. You could easily do the second method for a non-rectangular shape; however, figuring out how to clamp the coordinates will be harder than for rectangles. I'm sure there's a library or something out there that should do that for you already... 

This is what I mean by "the other half will be trimming the out-of-bounds coordinate back to the bounds of the defined polyline (trivial for a rectangle, more difficult for non-standard shapes)." Does this help clarify what I said above?

**Reply by SeanPluemer** [reply]
Yes, that does clarify it! I think I have what I need to move forward!

Thanks
Sean

**Comment by gursel58** [comment]
Just an additional question on using a custom size table. I am just getting together parts for a table and am wondering if there is a formula on how long to make the V channels ? And as far as placement of the v-channel is there a specific distance from the sides of the table ?
Thanks
Greg

**Reply by rdudhagra** [reply]
Hey Greg,
The lengths of the two side V channels should be identical. The back channel just needs to be 25mm shorter than the middle channel (the one that will hold the magnet carriage). You can put them as close to the sides of the tables as you want—just make sure there is enough clearance for the 3D-printed parts (and the heads of the M5 bolts that protrude out from them).

Let me know if you have any more questions!
